### PR TITLE
Implement `OrdWithEngines` for some types

### PR DESCRIPTION
**Pull request by emilyaherbert**
## Description

This PR is a subset of #3744. It adds a `OrdWithEngines` for the types required in #3744. Note---this isn't a necessary change into master right at this current moment, just simply a way of breaking up #3744 😄

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: sway-core/src/engine_threading.rs [hunk 1]
```diff
@@ -151,8 +169,19 @@ impl<T: PartialEqWithEngines + ?Sized> PartialEqWithEngines for &T {
...
+}
+
+impl<T: OrdWithEngines> OrdWithEngines for Option<T> {
+    fn cmp(&self, other: &Self, type_engine: &TypeEngine) -> Ordering {
+        match (self, other) {
+            (Some(x), Some(y)) => x.cmp(y, type_engine),
+            (Some(_), None) => Ordering::Less,
+            (None, Some(_)) => Ordering::Greater,
+            (None, None) => Ordering::Equal,
+        }
```
Comment (otrho) [inline sway-core/src/engine_threading.rs:184]: This is interesting.  Saying `None` is greater than `Some` and that `None` can equal `None` isn't necessarily intuitive.  But if it's what fits, then... sure.  I'm kinda thinking `T` should be `PartialEq` and this delegates more.

**Review by otrho — APPROVED** [review]
Do you need `Ord` to put stuff into `BTreeMap`s or something?  What's the general need for ordering types?

**Comment by emilyaherbert** [comment]
> Do you need Ord to put stuff into BTreeMaps or something? What's the general need for ordering types?

Yeah, that is the reason!

File: sway-core/src/engine_threading.rs (hunk 1 above)
Comment (emilyaherbert) [reply sway-core/src/engine_threading.rs:184]: I got this implementation from this description here: https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#derivable

> When derived on enums, variants are ordered by their discriminants. By default, the discriminant is smallest for variants at the top, and largest for variants at the bottom.

**Pull request merged**